0000-deb-repo-sync-gpg-check.md

```diff
@@ -0,0 +1,178 @@
+rhnContentSourceGpg
+====================
+content_source_id   number not null foreign key rhnContentSource(id)
+gpg_key_id          number not null foreign key rhnCryptoKey(id)
+created             timestamp
+modified            timestamp
+
+```
+
+## Repo signature checking
+
+The GPG verification should be optional but a warning should be shown to the the user in case he's not providing information about the GPG key used to verify the metadata.
+
+### Repo metadata location and signatures
+
+_Note: this section is only informative_
+
+Debian/Ubuntu repos use GPG signing to ensure the integrity of the `Release` file. The signature can be either in a separate file `Release.gpg` or inline in a file called `InRelease`. If both files are present the `InRelease` file is preferred.
+
+Typically the structure of a repo is like this:
+```
+http://mirror.example.com/ubuntu  <- $ARCHIVE_ROOT
+|
++- dists/bionic                   <- Suite or Codename
+   |
+   +- main                        <- $COMPNENT directories
+      |
+      +- binary-amd64
+         |
+         +- Packages.gz           <- Packages indices
+         +- Packages.xz
+         . . .
+      +- binary-i386
+      . . .
+   +- universe
+   . . .
+   +- Release                     <- Plain file
+   +- InRelease                   <- Release file + inline signature
+   +- Release.gpg                 <- Separate signature
+
+```
+
+E.g. for official Ubuntu repos the URL of these files:
+```
+http://mirror.example.com/ubuntu/dists/bionic-updates/InRelease
+http://mirror.example.com/ubuntu/dists/bionic-updates/Release.gpg
+```
+
+There's also an alternative layout for Debian repos called "flat repos". In this case the repo directory containins both `Release*` and `Packages*` files. E.g.:
+```
+https://download.opensuse.org/repositories/systemsmanagement:/Uyuni:/Master:/Ubuntu1804-Uyuni-Client-Tools/xUbuntu_18.04/
+```
+
+_Note #1:_
+
+In Uyuni the URL of a deb repository must point to the directory containing the binary packages for a specific architecture. E.g. for Ubuntu 18.04 updates repo is something like:
+```
+http://mirror.example.com/ubuntu/dists/bionic-updates/main/binary-amd64/
+```
+or in case of a flat repository to the directory containing both `Release*` and `Packages*` files:
+```
+https://download.opensuse.org/repositories/systemsmanagement:/Uyuni:/Master:/Ubuntu1804-Uyuni-Client-Tools/xUbuntu_18.04/
+```
+
+_Note #2:_
+
+Debian repos don't have a standard location for the public GPG keys used to verify the metadata. Sometimes the key is located in the root of a flat repo but the location is not standardized.
+
+### GPG keys management
+
+Uyuni ships the SUSE and OpenSUSE keys in the package `uyuni-build-keys`. The keys for Ubuntu and Debian should be shipped in this package too.
+
+Any additional keys that need to be used (e.g. for Ubuntu derivatives distros or for custom repos) must be created manually using the existing UI (`Systems -> Autoinstallation -> GPG and SSL keys`)
```

Comment:
spacewalk-repo-sync ask to trust the key and if the user answer with "yes" it imports it.
For Debian/Ubuntu the key may not exist at a specific place so we have a problem to download / guess it.
But maybe we could ask a keyserver?

`Systems -> Autoinstallation -> GPG and SSL keys` is a place where Keys gets stored for autoinstallation.
Sometimes they are used also for different use cases, but it is not very intuitive .

Reply:
Note that [`distribution-gpg-keys`](https://github.com/xsuchy/distribution-gpg-keys) package in Fedora and openSUSE ships many of the RPM distribution GPG keys.

In Fedora, we ship the Debian keys in `debian-keyring` package. The Ubuntu keys are shipped in the `ubu-keyring` package.

We may want to standardize on this approach and unify all the keys we ship this way...

Reply:
Please see also my previous [comment](https://github.com/uyuni-project/uyuni-rfc/pull/12#discussion_r322756074). 

I think there could be several options for adding new GPG keys for third party repos:

1) The UI (`Systems -> Autoinstallation -> GPG and SSL keys`). Not very intuitive. Maybe we could move it to some other place in the menu.

2) Let the user fiddle directly with `/var/lib/spacewalk/gpgdir/pubring.gpg`. I think this is risk.

3) Write a small shell script to let the user import or remove third party gpg keys into `/var/lib/spacewalk/gpgdir/pubring.gpg`.